pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,332 @@
+            by default the tapes will be executed in serial.
+        execute_fn (callable): The execution function used to execute the tapes
+            during the forward pass. This function must return a tuple ``(results, jacobians)``.
+            If ``jacobians`` is an empty list, then ``gradient_fn`` is used to
+            compute the gradients during the backwards pass.
+        gradient_kwargs (dict): dictionary of keyword arguments to pass when
+            determining the gradients of tapes
+        gradient_fn (callable): the gradient function to use to compute quantum gradients
+        _n (int): a positive integer used to track nesting of derivatives, for example
+            if the nth-order derivative is requested.
+        max_diff (int): If ``gradient_fn`` is a gradient transform, this option specifies
+            the maximum order of derivatives to support. Increasing this value allows
+            for higher order derivatives to be extracted, at the cost of additional
+            (classical) computational overhead during the backwards pass.
+        mode (str): Whether the gradients should be computed on the forward
+            pass (``forward``) or the backward pass (``backward``).
+
+    Returns:
+        list[list[float]]: A nested list of tape results. Each element in
+        the returned list corresponds in order to the provided tapes.
+    """
+    # pylint: disable=unused-argument
+    if max_diff > 1:
+        raise ValueError("The JAX interface only supports first order derivatives.")
+
+    _validate_tapes(tapes)
+
+    for tape in tapes:
+        # set the trainable parameters
+        params = tape.get_parameters(trainable_only=False)
+        tape.trainable_params = qml.math.get_trainable_indices(params)
+
+    parameters = tuple(list(t.get_parameters()) for t in tapes)
+
+    if gradient_fn is None:
+        return _execute_with_fwd(
+            parameters,
+            tapes=tapes,
+            device=device,
+            execute_fn=execute_fn,
+            gradient_kwargs=gradient_kwargs,
+            _n=_n,
+        )
+
+    return _execute(
+        parameters,
+        tapes=tapes,
+        device=device,
+        execute_fn=execute_fn,
+        gradient_fn=gradient_fn,
+        gradient_kwargs=gradient_kwargs,
+        _n=_n,
+    )
+
+
+def _validate_tapes(tapes):
+    """Validates that the input tapes are compatible with JAX support.
+
+    Note: the goal of this validation is to filter out cases where ragged
+    outputs for QNodes may arise. Such QNodes involve creating arrays from
+    ragged nested sequences that can not be handled by JAX.
+
+    Raises:
+        ValueError: if tapes that produce ragged outputs were provided
+    """
+    for t in tapes:
+
+        return_types = [o.return_type for o in t.observables]
+        set_of_return_types = set(return_types)
+        probs_or_sample_measure = Sample in return_types or Probability in return_types
```

Comment:
I wonder if this could be made slightly more efficient by doing
```suggestion
        probs_or_sample_measure = any(o.return_type in (Sample, Probability) for o in t.observables)
```

since the iteration will stop as soon as a single `True` is encountered. It will also avoid the overhead of:

- converting a list into a set
- Checking the set for `Sample`
- Checking the set of `Probability`